### PR TITLE
Validating PropTypes when a class is created

### PR DESCRIPTION
**Pull request by rppc**
**Note:** [This](https://github.com/facebook/react/pull/5475) is the same pull request. I've cleaned up the commit history.

Following [this](https://github.com/facebook/react/issues/5468) bug report, the PropTypes are now checked in development mode when the class is created, so that, for instance, the call:

```javascript
var MyComponent = React.createClass({
        displayName: 'Component',
        propTypes: {
            optionalNumber: React.PropTypes.number,
            optionalEnum: React.PropTypes.oneOf('foo'),       // Invalid.
          },
          render: function() {
            return <span>{this.props.prop}</span>;
          },
        });
```
will result in a warning being logged, and *optionalEnum* being cleared from the **propTypes** object.

Up to this point, this check only occurred when **ReactElement.createElement** was called in development mode. Some refactoring might be needed in order to avoid unnecessary tests when **ReactElement.createElement** is called in development mode.

Test cases were included in the file [ReactClass-test.js](https://github.com/rppc/react/blob/changed/src/isomorphic/classic/class/__tests__/ReactClass-test.js#L384) (see the last test).

### CODE REVIEW

**Comment by phss** [comment]
This is a bit of a newbie question but why not validate for ContextType as well? As far as I understand, they are defined in the same way as PropTypes. Shouldn't they also raise a warning if defined in an invalid way?

**Comment by rppc** [comment]
You're probably right. My knowledge of React is very limited, as this is my first contribution ever.

**Comment by zpao** [comment]
@phss That's a point I was going to bring up while reviewing :)

I appreciate you taking the time to do this and I know it was a pain to work through git issues but I have a fundamental change to make.

We should simplify this and just put some `if (__DEV__) { if (invalidPropType) { warning(…) } }` blocks inside ReactPropTypes. ReactClass shouldn't need to be involved at all. All tests would stay with PropTypes.

Basically, if I were to call `React.PropTypes.oneOf('foo')` outside of any React component, I should see a warning. This would ensure that contextTypes would also warn.

File: src/isomorphic/classic/types/ReactPropTypes.js [hunk 1]
```diff
@@ -208,11 +210,13 @@ function createInstanceTypeChecker(expectedClass) {
 
 function createEnumTypeChecker(expectedValues) {
   if (!Array.isArray(expectedValues)) {
-    return createChainableTypeChecker(function() {
+    var error = createChainableTypeChecker(function() {
```
Comment (zpao) [inline src/isomorphic/classic/types/ReactPropTypes.js]: Since this is the example I keep using, let's work through it.

1. We put `if (__DEV__) { warning(false, 'Some message here') }
2. Continue returning a chainable type checker (because isRequired can still be added).
3. We can perhaps just make it so that we pass it an empty function in or we keep it returning an error but we change that message to make it clear that this prop failure can be ignored if you are not the component author. I'm leaning towards the former right this second (because it's not terribly useful to have to work past a failure you aren't responsible for).

File: src/isomorphic/classic/class/__tests__/ReactClass-test.js [hunk 2]
```diff
@@ -381,4 +381,30 @@ describe('ReactClass-spec', function() {
     );
   });
 
+  it('should remove invalid prop types', function() {
+    var warn = console.error;
+    console.error = jest.genMockFn();
```
Comment (zpao) [inline src/isomorphic/classic/class/__tests__/ReactClass-test.js]: Don't try catch this. Use spyOn (we do this a bunch and already do it inside the proptypes test)

**Comment by rppc** [comment]
@zpao Thanks for the review!

Like I said, this is my first contribution to React, so I apologise if I missed some important details.

What you said made sense. I will try to look into it when I have the chance.

File: src/isomorphic/classic/types/ReactPropTypes.js (hunk 1 above)
Comment (rppc) [inline src/isomorphic/classic/types/ReactPropTypes.js]: I think I get the idea!

File: src/isomorphic/classic/class/__tests__/ReactClass-test.js (hunk 2 above)
Comment (rppc) [inline src/isomorphic/classic/class/__tests__/ReactClass-test.js]: Got it!

**Comment by rppc** [comment]
Is this what you had in mind? If so, I did over-complicate things a bit...

I'm so sorry... I accidentally merged all those unnecessary commits back in... Please tell me if should clean them up and I'll go on and just create a new pull request.

Two unit tests are failing because they expect the calls to ReactPropTypes functions not to warn. I've decided not to change those just yet.

**Comment by zpao** [comment]
Please don't make another PR, just clean up your branch (strip that merge commit and don't run `git merge master` again (and/or fix your master branch)). It's important to use a singe PR so that we can track review comments and discussions. Looking across 10 PRs for that is super difficult, especially when you do it a year or 2 from now.

Please use the local environment to run tests and don't use Travis CI. Since we have a comment come in for every change, it gets noisy. `npm test` will run the tests.

Otherwise yes, this is on the right track for what I meant.

**Comment by rppc** [comment]
Yes, I shouldn't have merged that master branch... Silly me! I think I've got it now!

I'll fix the tests that are failing as soon as I can.

**Comment by rppc** [comment]
@zpao As promised, I fixed the tests that were expecting functions **oneOf()** and **oneOfType()** not to warn and also some of the code so that all tests would pass.

File: src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js [hunk 3]
```diff
@@ -569,11 +569,19 @@ describe('ReactPropTypes', function() {
 
   describe('OneOf Types', function() {
     it('should fail for invalid argument', function() {
+      spyOn(console, 'error');
+      var warn = console.error;
+      console.error = jest.genMockFn();
```
Comment (zpao) [inline src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js]: No need to do this, spies are mock functions. And no need to assign to a temporary variable and back - spies are cleaned up automatically. Here's a complete & short example usage: https://github.com/facebook/react/blob/master/src/addons/__tests__/ReactFragment-test.js#L93-L100

File: src/isomorphic/classic/types/ReactPropTypes.js [hunk 4]
```diff
@@ -13,6 +13,7 @@
 
 var ReactElement = require('ReactElement');
 var ReactPropTypeLocationNames = require('ReactPropTypeLocationNames');
+var warning = require('warning');
```
Comment (zpao) [inline src/isomorphic/classic/types/ReactPropTypes.js]: Please move this require under `getIteratorFn` (style isn't totally obvious here, sorry)

File: src/isomorphic/classic/class/__tests__/ReactClass-test.js [hunk 5]
```diff
@@ -381,4 +381,27 @@ describe('ReactClass-spec', function() {
...
+    });
+    expect(console.error.mock.calls.length).toBe(2);
+    expect(console.error.mock.calls[0][0]).toContain(
+        'Invalid argument supplied to oneOf, expected an instance of array.'
+    );
+    expect(console.error.mock.calls[1][0]).toContain(
+        'Invalid argument supplied to oneOfType, expected an instance of array.'
+    );
+    console.error = warn;
+  });
```
Comment (zpao) [inline src/isomorphic/classic/class/__tests__/ReactClass-test.js]: Let's cleanup the mocking here as I described earlier.

**Comment by rppc** [comment]
@zpao Done!

**Comment by jimfb** [comment]
Ping @zpao

**Comment by gaearon** [comment]
Hey @rppc! We are sorry this got under our radar and doesn’t merge cleanly now. Someone else took a stab at it in #6316 which looks fairly similar so we’ll probably go ahead with it. Thank you for taking the time to contribute, and sorry for spending your time on the back-and-forth without merging the result.

**Comment by rppc** [comment]
@gaearon That's OK!
